File: zero-server/src/main/kotlin/com/zeroday/handler/handlers/ChallengeHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.model.NotificationType
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import com.zeroday.service.ChallengeService
import com.zeroday.service.NotificationService
import kotlinx.serialization.json.JsonObject

class ChallengeHandler(
    private val challengeService: ChallengeService,
    private val notificationService: NotificationService
) : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.GET_CHALLENGES,
        RequestTypes.CLAIM_CHALLENGE,
        RequestTypes.ROTATE_CHALLENGES
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.GET_CHALLENGES -> list(ctx, playerId)
            RequestTypes.CLAIM_CHALLENGE -> claim(ctx, playerId, payload)
            RequestTypes.ROTATE_CHALLENGES -> rotate(ctx, playerId)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported challenge request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun list(ctx: HandlerContext, playerId: String): HandlerResult {
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val now = System.currentTimeMillis()
        val rows = player.activeChallenges.map { c ->
            val def = challengeService.get(c.challengeId) ?: return@map mapOf("id" to c.challengeId)
            mapOf(
                "id" to def.id,
                "name" to def.name,
                "description" to def.description,
                "cadence" to def.cadence.name,
                "target" to def.targetValue,
                "current" to c.currentValue,
                "percent" to c.percent(def.targetValue),
                "completed" to c.completed,
                "claimed" to c.claimed,
                "expires_at" to c.expiresAt,
                "expires_in_ms" to (c.expiresAt - now).coerceAtLeast(0L),
                "rewards" to mapOf(
                    "credits" to def.rewardCredits,
                    "xp" to def.rewardXp,
                    "reputation" to def.rewardReputation
                )
            )
        }
        return HandlerResult.Ok(
            type = ResponseTypes.CHALLENGES,
            payload = mapOf(
                "challenges" to rows,
                "last_rotation" to player.lastChallengeRotation
            )
        )
    }

    private suspend fun claim(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val challengeId = try { payload.reqStr("challenge_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Challenge ID required", ErrorCategory.VALIDATION))
        }
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val active = player.activeChallenges.firstOrNull { it.challengeId == challengeId }
            ?: return HandlerResult.Err(ServerError.fromMessage("Unknown active challenge: $challengeId", ErrorCategory.NOT_FOUND))
        if (!active.completed) {
            return HandlerResult.Err(ServerError.fromMessage("Challenge not yet completed", ErrorCategory.VALIDATION))
        }
        if (active.claimed) {
            return HandlerResult.Ok(
                type = ResponseTypes.CHALLENGE_COMPLETED,
                payload = mapOf("challenge_id" to challengeId, "already_claimed" to true, "rewards" to emptyMap<String, Any>())
            )
        }
        val def = challengeService.get(challengeId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Unknown challenge definition: $challengeId", ErrorCategory.NOT_FOUND))
        active.claimed = true
        ctx.services.playerService.addCredits(playerId, def.rewardCredits)
        ctx.services.playerService.addReputation(playerId, def.rewardReputation)
        ctx.services.playerService.addExperience(playerId, def.rewardXp)
        notificationService.push(
            player = player,
            type = NotificationType.REWARD_GRANTED,
            title = "Challenge complete: ${def.name}",
            message = "You earned ${def.rewardCredits} credits and ${def.rewardXp} XP.",
            data = mapOf("challenge_id" to def.id, "credits" to def.rewardCredits.toString())
        )
        return HandlerResult.Ok(
            type = ResponseTypes.CHALLENGE_COMPLETED,
            payload = mapOf(
                "challenge_id" to challengeId,
                "rewards" to mapOf(
                    "credits" to def.rewardCredits,
                    "xp" to def.rewardXp,
                    "reputation" to def.rewardReputation
                )
            )
        )
    }

    private suspend fun rotate(ctx: HandlerContext, playerId: String): HandlerResult {
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        challengeService.forceRotate(player)
        return HandlerResult.Ok(
            type = ResponseTypes.CHALLENGE_ROTATED,
            payload = mapOf("challenges" to player.activeChallenges.map { it.challengeId })
        )
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/ChallengeHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;
using ZeroDayMMO.Shared.Models;

namespace ZeroDayMMO.Server.Handlers;

public class ChallengeHandler : IHandler
{
    private readonly IPlayerService _playerService;
    private readonly ChallengeService _challengeService;
    private readonly INotificationService _notificationService;

    public string MessageType => "get_challenges";

    public ChallengeHandler(
        IPlayerService playerService,
        ChallengeService challengeService,
        INotificationService notificationService)
    {
        _playerService = playerService;
        _challengeService = challengeService;
        _notificationService = notificationService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return new MessageResult("error", connectionId, new { message = "Missing payload" });

        var type = payload.Value.Str("type") ?? "get_challenges";

        return type switch
        {
            "get_challenges" => await List(connectionId),
            "claim_challenge" => await Claim(connectionId, payload.Value),
            "rotate_challenges" => await Rotate(connectionId),
            _ => new MessageResult("error", connectionId, new { message = $"Unsupported challenge request: {type}" })
        };
    }

    private async Task<IActionResult> List(string connectionId)
    {
        var player = _playerService.GetPlayer(connectionId);
        if (player is null)
            return new MessageResult("error", connectionId, new { message = "Player not found" });

        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        var rows = player.ActiveChallenges
            .Where(c => _challengeService.Get(c.ChallengeId) != null)
            .Select(c =>
        {
            var def = _challengeService.Get(c.ChallengeId)!;
            return new
            {
                id = c.ChallengeId,
                name = def.Name,
                description = def.Description,
                cadence = def.Cadence.ToString(),
                target = def.TargetValue,
                current = c.CurrentValue,
                percent = def.TargetValue == 0 ? 100 : (int)Math.Clamp(c.CurrentValue * 100 / def.TargetValue, 0, 100),
                completed = c.Completed,
                claimed = c.Claimed,
                expires_at = c.ExpiresAt,
                expires_in_ms = Math.Max(0, c.ExpiresAt - now),
                rewards = new
                {
                    credits = def.RewardCredits,
                    xp = def.RewardXp,
                    reputation = def.RewardReputation
                }
            };
        }).ToList();

        return new MessageResult("challenges", connectionId, new
        {
            challenges = rows,
            last_rotation = player.LastChallengeRotation
        });
    }

    private async Task<IActionResult> Claim(string connectionId, JsonElement payload)
    {
        string challengeId;
        try
        {
            challengeId = payload.ReqStr("challenge_id");
        }
        catch
        {
            return new MessageResult("error", connectionId, new { message = "Challenge ID required" });
        }

        var player = _playerService.GetPlayer(connectionId);
        if (player is null)
            return new MessageResult("error", connectionId, new { message = "Player not found" });

        var active = player.ActiveChallenges.FirstOrDefault(c => c.ChallengeId == challengeId);
        if (active is null)
            return new MessageResult("error", connectionId, new { message = $"Unknown active challenge: {challengeId}" });

        if (!active.Completed)
            return new MessageResult("error", connectionId, new { message = "Challenge not yet completed" });

        if (active.Claimed)
        {
            return new MessageResult("challenge_completed", connectionId, new
            {
                challenge_id = challengeId,
                already_claimed = true,
                rewards = new { }
            });
        }

        var def = _challengeService.Get(challengeId);
        if (def is null)
            return new MessageResult("error", connectionId, new { message = $"Unknown challenge definition: {challengeId}" });

        active.Claimed = true;

        await _playerService.AddCreditsAsync(connectionId, def.RewardCredits);
        await _playerService.AddReputationAsync(connectionId, def.RewardReputation);
        await _playerService.AddExperienceAsync(connectionId, def.RewardXp);

        _notificationService.AddNotification(connectionId, new Notification
        {
            Type = NotificationType.REWARD_GRANTED,
            Title = $"Challenge complete: {def.Name}",
            Message = $"You earned {def.RewardCredits} credits and {def.RewardXp} XP.",
            Data = new Dictionary<string, string>
            {
                { "challenge_id", def.Id },
                { "credits", def.RewardCredits.ToString() }
            }
        });

        return new MessageResult("challenge_completed", connectionId, new
        {
            challenge_id = challengeId,
            rewards = new
            {
                credits = def.RewardCredits,
                xp = def.RewardXp,
                reputation = def.RewardReputation
            }
        });
    }

    private async Task<IActionResult> Rotate(string connectionId)
    {
        var player = _playerService.GetPlayer(connectionId);
        if (player is null)
            return new MessageResult("error", connectionId, new { message = "Player not found" });

        await _challengeService.ForceRotate(player);

        return new MessageResult("challenge_rotated", connectionId, new
        {
            challenges = player.ActiveChallenges.Select(c => c.ChallengeId).ToList()
        });
    }
}
